### PR TITLE
Liftn functions for Option, Result and Validation

### PR DESCRIPTION
**Pull request by stephentetley**
Hi Magnus 

Here are the lifters for Option, Result and Validation. Plus `toErr` etc. on Option.

### CODE REVIEW

**Comment by stephentetley** [comment]
Oh, I though I'd fixed the current bug - I'll check to make sure I've pushed everything...

**Comment by stephentetley** [comment]
I must have missed committing a fix for that build error in the `TestValidation.flix` it has now been pushed.

File: main/src/library/Option.flix [hunk 1]
```diff
@@ -214,6 +214,38 @@ namespace Option {
         case Some((k, v))   => Map.singleton(k, v)
     }
 
+    ///
+    /// Returns the Option value `Ok(v) if `o` is `Some(v)`. Otherwise returns `Err(e)`.
```
Comment (magnus-madsen) [inline main/src/library/Option.flix:218]: Missing a backtick on `Ok(v)`.

File: main/src/library/Option.flix [hunk 2]
```diff
@@ -214,6 +214,38 @@ namespace Option {
...
+    ///
+    pub def toOk(o: Option[t], e: e): Result[t, e] = match o {
+        case None       => Err(e)
+        case Some(a)    => Ok(a)
+    }
+
+    ///
+    /// Returns the Option value `Err(e) if `o` is `Some(e)`. Otherwise returns `Ok(d)`.
```
Comment (magnus-madsen) [inline main/src/library/Option.flix:226]: Missing backtick.

File: main/src/library/Option.flix [hunk 3]
```diff
@@ -214,6 +214,38 @@ namespace Option {
...
+    ///
+    pub def toErr(o: Option[e], d: t): Result[t, e] = match o {
+        case None       => Ok(d)
+        case Some(e)    => Err(e)
+    }
+
+    ///
+    /// Returns the Validation value `Success(v) if `o` is `Some(v)`. Otherwise lifts `e` into Validation's `Failure`.
```
Comment (magnus-madsen) [inline main/src/library/Option.flix:234]: Missing backtick.

File: main/src/library/Result.flix [hunk 4]
```diff
@@ -183,6 +183,89 @@ namespace Result {
...
     }
 
+    ///
+    /// Applies the function in `r1` to the value in `r2`.
+    ///
+    pub def ap(r1: Result[t -> u & f, e], r2: Result[t, e]): Result[u, e] & f = match r1 {
+        case Err(e)     => Err(e)
+        case Ok(f)      => match r2 {
+            case Err(e)     => Err(e)
+            case Ok(a)    => Ok(f(a))
```
Comment (magnus-madsen) [inline main/src/library/Result.flix:193]: - Indention issue.
- Usually we have the Ok case before the Err case.

**Comment by stephentetley** [comment]
Thanks Magnus - I've just update to upstream/master and will be pushing those changes.

`TestValidation.flix` has the tests you wrote indented with one level and the tests I wrote with no indentation as they were very long lines.

Do you want me to unify them to one way or the other?

**Comment by magnus-madsen** [comment]
> TestValidation.flix has the tests you wrote indented with one level and the tests I wrote with no indentation as they were very long lines.

I have no strong feelings, so I am just going to merge it as is.

**Comment by stephentetley** [comment]
Thanks

**Pull request merged**